### PR TITLE
U4 10147 - Bypass User Security option in pickers

### PR DESCRIPTION
**Pull request by dnwhte**
This addresses the feature request at http://issues.umbraco.org/issue/U4-10147.

List of pickers supported:

Content Picker List View currently has an issue. Umbraco's user permissions do not work on the content picker list view ATM. http://issues.umbraco.org/issue/U4-10876

**Update:**
Issue preventing content picker list view from working is https://issues.umbraco.org/issue/U4-11488.

### CODE REVIEW

**Comment by emmaburstow** [comment]
Hi @danwhite85 

Thanks for this PR! I've already started testing and am really pleased with what it does. I'll have a play this evening and let you know if there's anything I need from you at all. It's a really useful change.

Emma, [Community Pull Request team](https://umbraco.com/blog/introducing-the-new-pr-team/)

**Comment by danwhite85** [comment]
I've created a new issue that relates to the Content Picker List View. The issue mentioned initially was only part of the problem apparently. 

http://issues.umbraco.org/issue/U4-11488

**Comment by danwhite85** [comment]
Anything I can do to get this moving? @emmaburstow @nul800sebastiaan

**Comment by emmaburstow** [comment]
Hi @danwhite85 

Sorry for the delay on this one. We're going to take a look at this next week and get back to you. The work on the security patch has been a priority and things have slowed here in the meantime. 

I promise we haven't forgotten you though! Speak soon,

Emma

**Comment by danwhite85** [comment]
Hi @emmaburstow @nul800sebastiaan-

Not trying to be a pain, but this PR has been hanging out in the Approved list for some time now (while others are being cycled). I know it's not a small PR and probably requires more review, but is there anything I do to expedite the process?

Thank you

Dan

**Comment by emmaburstow** [comment]
Hi @danwhite85 

We have actually had a good chat about this so please don't worry that we've forgotten about you! I'll prod him again and give you some feedback in the next few days.

Emma

**Comment by nul800sebastiaan** [comment]
Thanks @danwhite85 - It's a bit of a large PR and I've been trying to make it fit somewhere but it hasn't really worked so far. I need some feedback from the team before we can merge this in and I'm trying to get someone to look at it in the next few weeks.

Many thanks so far for the effort, this looks like a very useful addition! 👍 ⭐️

**Comment by danwhite85** [comment]
@nul800sebastiaan Thank you for the response. I realize it's a difficult PR to fit into a sprint and appreciate your trying to fit it in. As v8 approaches I am getting concerned about about the staleness of the PR and the ability to merge it in if it's not completed by time v8 launches. 

Would it be better if I created a separate PR for each picker type? Or should I resubmit it as a PR for the v8 branch?

Thank you again.

**Comment by dawoe** [comment]
Hi @danwhite85 

Sebastiaan is on holiday at the moment. I will assign him here. And I hope I remember to remind him about this one once he is back.

Dave [Community Pull Request team](https://umbraco.com/blog/introducing-the-new-cpr-team/)

**Comment by dnwhte** [comment]
@nul800sebastiaan not to be a pain, but any thoughts on this?
> @nul800sebastiaan Thank you for the response. I realize it's a difficult PR to fit into a sprint and appreciate your trying to fit it in. As v8 approaches I am getting concerned about about the staleness of the PR and the ability to merge it in if it's not completed by time v8 launches.
> 
> Would it be better if I created a separate PR for each picker type? Or should I resubmit it as a PR for the v8 branch?
> 
> Thank you again.

**Comment by nul800sebastiaan** [comment]
Some small notes on my test results so far: 

- The link picker in the rich text editor gives me a `403 forbidden` when selecting a link
image
- This will also need to be applied to Multi URL Picker which is now part of the core
- We will have to merge this up to v8

**Comment by dnwhte** [comment]
Thanks, @nul800sebastiaan. 

Could you clarify "We will have to merge this up to v8"? Does that mean getting this into v7 is not an option and the PR should be submitted against v8 instead?

**Comment by nul800sebastiaan** [comment]
@dnwhte 

What I mean is: everything we add to v7 will also need to be merged up to v8, we are constantly merging from 7 up to 8. Unfortunately these changes that affect loads of files that already are very different in v8, so it's.. not for the faint of heart (I just tried it and .. yikes! 😱 )

Anyway, it's still great functionality and we'd love to have it!

- Do you have some time to add support for Multi URL Picker?
- Do you have some time to see what went wrong with the RTE link picker?

**Comment by dnwhte** [comment]
Gotcha. Will look into it tomorrow.

**Comment by nul800sebastiaan** [comment]
Thanks @dnwhte - you're a star! 🥇

**Comment by dnwhte** [comment]
@nul800sebastiaan the 403 issue is resolved and support added for the Multi URL Picker. Anything else you need from me to move forward?

**Comment by nul800sebastiaan** [comment]
Cool, now we need to get it tested etc.

I just realized we also need to check if this works with Courier and Deploy. I'll get back to you when I've given it another good spin, thanks very much so far! 👍 👍

**Comment by Shazwazza** [comment]
It's mentioned at the top of this issue that the content picker list view has an issue but this pr was merged to solve that apparently https://github.com/umbraco/Umbraco-CMS/pull/2486

**Comment by Shazwazza** [comment]
I also think the text should be changed since it's not quite accurate "Bypass user permissions", which also sounds quite scary. 

Perhaps "Ignore user start nodes" ? and the description should explain more of what this actually means, from the current description people won't really know why this option exists. Example "Selecting this option allows a user to choose nodes that they normally don't have access to"

**Comment by dnwhte** [comment]
Thanks for the feedback, @Shazwazza. I will check on the content picker list view issue. 

I'll do some thinking on the verbiage and get back to you.

**Comment by dnwhte** [comment]
@Shazwazza I just checked the content picker list view. Looks like it is still an issue. The details are reported at https://issues.umbraco.org/issue/U4-11488.

**Comment by dnwhte** [comment]
I'm good with the recommended verbiage changes. Any thoughts @nul800sebastiaan?

**Comment by nul800sebastiaan** [comment]
Verbiage changed and I've redone all the testing, I think it all works as expected now. Which means... 


PARTY! 🎉 🍰 🍻 🍾 

Merged and ready to go in 7.15. 👍 🏅 

Now I will have to do a v8 merge, that's going to be fun. 😉

Thanks so much @dnwhte for your patience and for the great job!

**Comment by dnwhte** [comment]
Woohoo! This is awesome. Thanks a lot for your time and work, @nul800sebastiaan!

**Comment by nul800sebastiaan** [comment]
This has now also been merged up to v8 🥵🥵😅

**Comment by emmaburstow** [comment]
Yay! Really chuffed about this one 👍

File: src/Umbraco.Web/Trees/ContentTreeControllerBase.cs [hunk 1]
```diff
@@ -190,7 +190,7 @@ private int GetTopNodeId(IUmbracoEntity entity)
 
         protected abstract UmbracoObjectTypes UmbracoObjectType { get; }
 
-        protected IEnumerable<IUmbracoEntity> GetChildEntities(string id)
+        protected IEnumerable<IUmbracoEntity> GetChildEntities(string id, FormDataCollection queryStrings)
```
[the remark this answers is not in the document]
Comment (nul800sebastiaan) [reply src/Umbraco.Web/Trees/ContentTreeControllerBase.cs]: Ah shoot, I fixed all the public methods but didn't notice the protected ones. Will have a look.

**Comment by dnwhte** [comment]
@Shazwazza just for reference. In Jan 2018 we vaguely discussed some of these things (querystrings, and providing on context to FilterAllowedOutgoingMediaAttribute) (https://issues.umbraco.org/issue/U4-10147#comment=67-43737). 

I agree with your concerns. Given your response in 2018, I'm not sure how to go about it without a querystring. It's not much better, but what if instead of passing "ignoreUserStartNodes" as a querystring we pass the picker id and check the picker's settings in the handler?

**Comment by Shazwazza** [comment]
Hi @dnwhte,

There's a bunch of code in this PR so I'm unsure how exactly it all is glue'd together but the main things that we cannot do is : bypass security on ContentController or MediaController. I'm unsure why these controllers are being requested for pickers, perhaps it's an issue with Umbraco in general, for example i can see in the code that mediaResource is being called for the mediaPicker controller by why? I mean surely we can get the children from the entityResource (this would also improve performance). Similarly contentResource is being called by linkPicker controller, but again why? This should also be using entityResource.  ... if neither of these is 'possible' then we should re-think how we can make this possible. 

Pickers should in theory only use EntityController. In this case the risk of exposing data that the user doesn't have access to is more minimal since an Entity doesn't contain much information ... well that's actually not entirely true in v7. In v7 and 8.0.x, The EntityController for media will return all media properties too which we cannot do. In 8.1 this will be fixed. 

In my opinion, to implement this while maintaining reasonable security we would need to:

### Trees

Don't use a query string like `ignoreUserStartNodes` and instead for pickers that are using the tree's, yes, please pass in the data type ID, then check if the tree is currently in dialog mode, then check if a data type id has been passed in, then lookup that data type, check if it is a type that has config like `ignoreUserStartNodes`, then check if that is allowed and proceed as necessary

### Entity data

Unfortunately this isn't going to be the most straightforward thing to do due to how the EntityController and EntityService work pre 8.1 since it returns more data than we should be allowed to return if bypassing security. So a decision will need to be made - do we postpone this feature for 8.1+ or do we proceed with the below requirements?

First we need to ensure that security cannot be bypassed at all on MediaController/ContentController. Then we need to make sure that mediaResource and contentResource are not used for anything with that changes required for this feature - such as linkPicker and mediaPicker and instead determine if entityResource can be used instead and if not then document why so we can figure out how to make that work.

If we are going to proceed with this feature in a 7.x version, then various things will need to be done:

* We will need to back port the EntityService changes done in 8.1 so that simple media can be looked up with it's media path (https://github.com/umbraco/Umbraco-CMS/issues/3498, https://github.com/umbraco/Umbraco-CMS/pull/5658) 
* We will need to add another method to the EntityController like GetMediaChildren which will use the new EntityService method ensuring that the bare minimum media data is returned and in turn update the entityResource and mediaPicker to use this endpoint. This endpoint will work like the trees (see above) where you can request children on behalf of a picker by optionally passing in the picker Id and doing the checks like in trees

### Result

The end result of this is that an authenticated user could still request data that they don't have access to given that they know a valid Data Type picker Id that has this option configured by making requests either to the content/media tree or to the `entityResource.getById` or the new `entityResource.getMediaChildren` (if we proceed with a 7.x version, else if we only target 8.1 then `entityResource.getChildren` will work). However the data returned will be a minimal subset of data for these entities which don't expose any property values. This also means that unless a Data Type is configured to bypass security there is no way that a user can simply request security to be bypassed.

**Comment by dnwhte** [comment]
Hi @Shazwazza -

I'm currently working on a v7 project for which this feature is required, so I'm willing to help however I can.

From what I can tell, the mediaPicker and linkPicker use the content/media resources in order to get extra properties that are not available with the entityResource. Media picker needs images and link picker needs the url and the properties array to pass to the tinymce service.

I'm not too familiar with the v8 codebase, but giving it a quick scan I'm assuming the general approach to do this would be to use entityResource and then supplement the data with mediaHelper and contentEditingHelper. Is this correct?

It looks like the media picker could go without the media resource by using [mediaHelper.resolveFileFromEntity](https://github.com/umbraco/Umbraco-CMS/blob/9a45bda430b30fa389229b444308039bd7d19f40/src/Umbraco.Web.UI.Client/src/common/services/mediahelper.service.js#L144) as seen [here](https://github.com/umbraco/Umbraco-CMS/blob/v7/dev/src/Umbraco.Web.UI.Client/src/views/common/overlays/mediaPicker/mediapicker.controller.js#L344-L373). 

In order to not use the contentResource in linkPicker more data would probably need to be added to metaData in the entityResource response or additional helpers in contentEditingHelper. The [linkPicker in v8 still uses the contentResource](https://github.com/umbraco/Umbraco-CMS/blob/342f924e8f044201247aabad98738966ac8bb80d/src/Umbraco.Web.UI.Client/src/views/common/infiniteeditors/linkpicker/linkpicker.controller.js#L149).

How would you like to proceed? I could start refactoring the ignoreUserStartNodes querystring to use the data type id instead.

Thanks again for your time.

**Comment by nul800sebastiaan** [comment]
Tagged as breaking since in in change https://github.com/umbraco/Umbraco-CMS/commit/966b07ba1c969934b322b0266a6a9e0e1abda015 `IUmbracoEntity GetByKey(Guid key, bool loadBaseType = true);` has been changed to `IUmbracoEntity GetByKey(Guid key);`.

**Comment by PeteDuncanson** [comment]
@nul800sebastiaan I'm a little confused by the last few comments on this one. Have @Shazwazza concerns about security be addressed/implemented or are they shipped as it in the latest v7.15 and v8.1 releases? What state is this now in?

**Comment by Shazwazza** [comment]
@PeteDuncanson yes the concerns were addressed and much had to be refactored. Many pickers had to be adjusted to use entityResource (which is what they should have been using anyways) and you cannot request data that you don't have access to unless you have a data type configured to allow that scenario, the data returned is minimal (i.e. entityResource results)

**Pull request merged**